pdfplumber/utils.py

```diff
@@ -205,86 +205,103 @@ def bbox_to_rect(bbox):
-        else:
-            min_key, max_key = ("top", "bottom") if vertical_ttb else ("bottom", "top")
+def cluster_line_chars(
+    chars, tolerance, keep_blank_chars=False, min_key="x0", max_key="x1", sort_asc=True
+):
+    get_text = itemgetter("text")
+
+    words = []
+    current_word = []
 
-        words = []
-        current_word = []
+    comp_fn = gt if sort_asc else lt
+    tol_fn = add if sort_asc else sub
 
-        asc_order = (upright and horizontal_ltr) or (not upright and vertical_ttb)
+    def sort_key(x):
+        return tol_fn(0, x[min_key])
 
-        comp_fn = gt if asc_order else lt
-        tol_fn = add if asc_order else sub
+    sorted_chars = sorted(chars, key=sort_key)
 
-        def sort_key(x):
-            return tol_fn(0, x[min_key])
+    for char in sorted_chars:
+        if not keep_blank_chars and get_text(char).isspace():
+            if len(current_word) > 0:
+                words.append(current_word)
+                current_word = []
+        elif len(current_word) == 0:
+            current_word.append(char)
+        else:
+            last_char = current_word[-1]
+            prev_pos = tol_fn(last_char[max_key], tolerance)
+            if comp_fn(char[min_key], prev_pos):
+                words.append(current_word)
+                current_word = []
+            current_word.append(char)
+
+    if len(current_word) > 0:
+        words.append(current_word)
 
-        sorted_chars = sorted(chars, key=sort_key)
+    return words
 
-        for char in sorted_chars:
-            if not keep_blank_chars and get_text(char).isspace():
-                if len(current_word) > 0:
-                    words.append(current_word)
-                    current_word = []
-                else:
-                    pass
-            elif len(current_word) == 0:
-                current_word.append(char)
-            else:
-                last_char = current_word[-1]
-                prev_pos = tol_fn(last_char[max_key], tolerance)
-                if comp_fn(char[min_key], prev_pos):
-                    words.append(current_word)
-                    current_word = []
-                current_word.append(char)
 
-        if len(current_word) > 0:
-            words.append(current_word)
+def extract_words(
+    chars,
+    x_tolerance=DEFAULT_X_TOLERANCE,
+    y_tolerance=DEFAULT_Y_TOLERANCE,
+    keep_blank_chars=False,
+    horizontal_ltr=True,  # Should words be read left-to-right?
+    vertical_ttb=True,  # Should vertical words be read top-to-bottom?
+    extra_attrs=[],
+):
 
-        return [process_word_chars(chars, upright) for chars in words]
+    x_tolerance = decimalize(x_tolerance)
+    y_tolerance = decimalize(y_tolerance)
 
-    chars_by_upright = {1: [], 0: []}
     words = []
-    for char in to_list(chars):
-        chars_by_upright[char.get("upright", 1)].append(char)
+    grouped = itertools.groupby(chars, itemgetter("upright", *extra_attrs))
```

Comment:
`itertools.groupby()` expects the iterable to be sorted. Should that matter here? I think so not.

Reply:
This small-but-smart question led me more broadly to question assumptions about sort-order in `.extract_words`, and led me to (a) rewriting the logic again, in a hopefully more robust and clearly-defined way, and (b) adding a `use_text_flow` parameter. (Open to ideas for other names for that.) More details below, in 84f9181.

Reply:
And, just to follow up on the original question: In order to make the order of the words returned by `.extract_words` predictable and consistent (top-down, and then either left-right or right-left, depending on `horizontal_ltr`), I don't think we want to sort by `extra_attrs`... but I'm open to other perspectives on this.